packages/ic-asset-certification/src/asset_router.rs

```diff
@@ -0,0 +1,719 @@
+///         headers: vec![(
+///             "cache-control".to_string(),
+///             "public, max-age=31536000, immutable".to_string(),
+///         )],
+///     },
+///     AssetConfig::Pattern {
+///         pattern: "**/*.css".to_string(),
+///         content_type: Some("text/css".to_string()),
+///         headers: vec![(
+///             "cache-control".to_string(),
+///             "public, max-age=31536000, immutable".to_string(),
+///         )],
+///     },
+/// ];
+///
+/// asset_router
+///     .certify_assets(&mut http_certification_tree, assets, asset_configs)
+///     .unwrap();
+///
+/// let index_html_request = HttpRequest {
+///     method: "GET".to_string(),
+///     url: "/".to_string(),
+///     headers: vec![],
+///     body: vec![],
+/// };
+///
+/// let (index_html_response, index_html_tree, index_html_expr_path) = asset_router
+///     .serve_asset(&http_certification_tree, &index_html_request)
+///     .unwrap();
+///
+/// ```
+#[derive(Debug)]
+pub struct AssetRouter<'a> {
+    responses: HashMap<String, CertifiedAssetResponse<'a>>,
+    fallback_responses: HashMap<String, CertifiedAssetResponse<'a>>,
+}
+
+const IC_CERTIFICATE_EXPRESSION_HEADER: &str = "IC-CertificateExpression";
+
+impl<'a> AssetRouter<'a> {
+    /// Creates a new [AssetRouter].
+    pub fn new() -> Self {
+        AssetRouter {
+            responses: HashMap::new(),
+            fallback_responses: HashMap::new(),
+        }
+    }
+
+    /// Returns the corresponding
+    /// [HttpResponse](ic_http_certification::HttpResponse) for the provided
+    /// [HttpRequest](ic_http_certification::HttpRequest) if it is found
+    /// in the router, along with the
+    /// [certification witness](ic_certification::HashTree) and the
+    /// corresponding
+    /// [expression path](ic_http_certification::HttpCertificationPath).
+    ///
+    /// If an exact match is not found, then a fallback will
+    /// be searched for. See the
+    /// [fallback_for](AssetConfig::File::fallback_for) configuration
+    /// option for more information on fallbacks.
+    ///
+    /// Returns [None] if no suitable
+    /// [HttpResponse](ic_http_certification::HttpResponse) is found for the
+    /// given [HttpRequest](ic_http_certification::HttpRequest).
+    pub fn serve_asset(
+        &self,
+        http_certification_tree: &HttpCertificationTree,
+        request: &HttpRequest,
+    ) -> AssetCertificationResult<(HttpResponse, HashTree, Vec<String>)> {
+        if let Some(CertifiedAssetResponse {
+            response,
+            tree_entry,
+        }) = self.responses.get(&request.url)
+        {
+            let witness = http_certification_tree.witness(tree_entry, &request.url)?;
+            let expr_path = tree_entry.path.to_expr_path();
+
+            return Ok((response.clone().into(), witness, expr_path));
+        }
+
+        let mut url_scopes = request.url.split('/').collect::<Vec<_>>();
```

Comment:
i would add a comment here to note that the request.url will actually just include the path and not the domain part of the url, otherwise this scope parsing will also include the protocol part which is not what is intended, maybe also add a check for that since it's not enforced in the HttpRequest type

Reply:
That's a good point, we're assuming the request is coming from a correct implementation of the HTTP Gateway protocol. I think adding the validation, just in case, wouldn't be a bad thing. Or adding some more flexible parsing. I've added a ticket for that: https://dfinity.atlassian.net/browse/TT-395